File: PlayerUI/Flutter/objc/PlayerFlutter/BCOVVideoPlayer.m
//
//  BCOVVideoPlayer.m
//  PlayerFlutter
//
//  Created by Carlos Ceja.
//

#import <AVFoundation/AVFoundation.h>
#import <AVKit/AVKit.h>

#import <BrightcovePlayerSDK/BrightcovePlayerSDK.h>

#import "BCOVVideoPlayer.h"


@interface BCOVVideoPlayer () <BCOVPlaybackControllerDelegate>

@end

@implementation BCOVVideoPlayer
{
    int64_t _viewId;
    
    AVPlayer *_player;
    AVPlayerViewController *_avpvc;
    
    FlutterEventSink _eventSink;
    
    FlutterEventChannel *_eventChannel;
    FlutterMethodChannel *_methodChannel;
    
    BCOVPlayerSDKManager *_manager;
    BCOVPlaybackService *_playbackService;
    id<BCOVPlaybackController> _playbackController;
}

- (instancetype)initWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args binaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenger
{
    if (self = [super init])
    {
        _viewId = viewId;
        
        _player = [[AVPlayer alloc] init];
        
        _avpvc = ({
            AVPlayerViewController *avpvc = [[AVPlayerViewController alloc] init];
            avpvc.view.frame = frame;
            avpvc.player = _player;
            avpvc.showsPlaybackControls = NO;
            avpvc;
        });
        
        [self setupEventChannelWithViewIdentifier:viewId messenger:messenger instance:self];
        [self setupMethodChannelWithViewIdentifier:viewId messenger:messenger instance:self];
        
        UIViewController *rootController = UIApplication.sharedApplication.delegate.window.rootViewController;
        [rootController addChildViewController:_avpvc];
        
        _manager = [BCOVPlayerSDKManager sharedManager];
        
        NSDictionary *parsedData = args;
        
        NSDictionary *playbackControllerArgs = parsedData[@"playbackController"];
        NSNumber *autoPlay = playbackControllerArgs[@"autoPlay"];
        NSNumber *autoAdvance = playbackControllerArgs[@"autoAdvance"];
        _playbackController = ({
            id<BCOVPlaybackController> controller = [_manager createPlaybackController];
            controller.delegate = self;
            controller.autoAdvance = autoAdvance.boolValue;
            controller.autoPlay = autoPlay.boolValue;
            controller.options = @{ kBCOVAVPlayerViewControllerCompatibilityKey: @(YES) };
            controller;
        });
        
        NSDictionary *playbackServiceArgs = parsedData[@"playbackService"];
        NSString *accountId = playbackServiceArgs[@"accountId"];
        NSString *policyKey = playbackServiceArgs[@"policyKey"];
        _playbackService = [[BCOVPlaybackService alloc] initWithAccountId:accountId policyKey:policyKey];
        
    }
    
    return self;
}

#pragma mark FlutterPlatformView

- (nonnull UIView *)view
{
    return _avpvc.view;
}

#pragma mark FlutterStreamHandler

- (FlutterError *)onListenWithArguments:(id)arguments eventSink:(FlutterEventSink)events
{
    _eventSink = events;
    return nil;
}

- (FlutterError *)onCancelWithArguments:(id)arguments
{
    _eventSink = nil;
    return nil;
}

#pragma mark BCOVPlaybackControllerDelegate

- (void)playbackController:(id<BCOVPlaybackController>)controller didAdvanceToPlaybackSession:(id<BCOVPlaybackSession>)session
{
    _player = session.player;
    _avpvc.player = session.player;
    
    id duration = session.video.properties[@"duration"];
    _eventSink(@{@"name": @"didAdvanceToPlaybackSession", @"duration": duration, @"isPlaying": @(self->_playbackController.isAutoPlay)});
}

- (void)playbackController:(id<BCOVPlaybackController>)controller playbackSession:(id<BCOVPlaybackSession>)session didProgressTo:(NSTimeInterval)progress
{
    if (!isinf(progress))
    {
        _eventSink(@{@"name": @"didProgressTo", @"progress":@(progress * 1000)});
    }
}

#pragma mark Private Methods

- (void)setupEventChannelWithViewIdentifier:(int64_t)viewId messenger:(NSObject<FlutterBinaryMessenger> *)messenger instance:(BCOVVideoPlayer *)instance
{
    // register for Flutter event channel
    NSString *eventChannelName = [NSString stringWithFormat:@"bcov.flutter/event_channel_%lld", viewId];
    instance->_eventChannel = [FlutterEventChannel eventChannelWithName:eventChannelName binaryMessenger:messenger codec:FlutterJSONMethodCodec.sharedInstance];
    [instance->_eventChannel setStreamHandler:instance];
}

- (void)setupMethodChannelWithViewIdentifier:(int64_t)viewId messenger:(NSObject<FlutterBinaryMessenger> *)messenger instance:(BCOVVideoPlayer *)instance
{
    // register for Flutter method channel
    NSString *methodChannelName = [NSString stringWithFormat:@"bcov.flutter/method_channel_%lld", viewId];
    instance->_methodChannel = [FlutterMethodChannel methodChannelWithName:methodChannelName binaryMessenger:messenger];
    
    __weak typeof(self) weakSelf = self;
    [instance->_methodChannel setMethodCallHandler:^(FlutterMethodCall * _Nonnull call, FlutterResult _Nonnull result)
     {
        [weakSelf handleMethodCall:call result:result];
    }];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result
{
    if ([@"setVideo" isEqualToString:call.method])
    {
        NSString *videoId = call.arguments[@"videoId"];
        NSString *authToken = ([[call.arguments[@"authToken"] class] isKindOfClass:[NSNull class]] ?
                               call.arguments[@"authToken"] :
                               nil);
        NSDictionary *parameters = ([[call.arguments[@"parameters"] class] isKindOfClass:[NSNull class]] ?
                                    call.arguments[@"parameters"] :
                                    nil);
        
        [_playbackService findVideoWithVideoID:videoId authToken:authToken parameters:parameters completion:^(BCOVVideo *video, NSDictionary *jsonResponse, NSError *error)
         {
            if (video)
            {
                [self->_playbackController setVideos:@[ video ]];
            }
            else
            {
                self->_eventSink(@{@"name": @"onError"});
            }
            
            result(@(YES));
        }];
    }
    else if ([@"play" isEqualToString:call.method])
    {
        [_playbackController play];
    }
    else if ([@"pause" isEqualToString:call.method])
    {
        [_playbackController pause];
    }
    else
    {
        result(FlutterMethodNotImplemented);
    }
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    _avpvc.player = nil;
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    _avpvc.player = _player;
}

@end


@implementation BCOVVideoPlayerFactory
{
    NSObject<FlutterPluginRegistrar> *_registrar;
    NSObject<FlutterBinaryMessenger> *_messenger;
}

- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar
{
    if (self = [super init])
    {
        _registrar = registrar;
        
        _messenger = registrar.messenger;
    }
    
    return self;
}

#pragma mark FlutterPlatformViewFactory

- (nonnull NSObject<FlutterPlatformView> *)createWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id _Nullable)args
{
    return [[BCOVVideoPlayer alloc] initWithFrame:frame viewIdentifier:viewId arguments:args binaryMessenger:_messenger];
}

- (NSObject<FlutterMessageCodec> *)createArgsCodec
{
    return FlutterStandardMessageCodec.sharedInstance;
}

@end

File: PlayerUI/Flutter/swift/PlayerFlutter/BCOVVideoPlayer.swift
//
//  BCOVVideoPlayer.swift
//  PlayerFlutter
//
//  Created by Carlos Ceja.
//

import AVFoundation
import AVKit
import Foundation

import BrightcovePlayerSDK
import Flutter


class BCOVVideoPlayer: NSObject {
    
    private var viewId: Int64
    
    lazy private var player: AVPlayer = {
        var _player = AVPlayer()
        return _player
    }()
    
    lazy private var avpvc: AVPlayerViewController = {
        var _avpvc = AVPlayerViewController()
        _avpvc.player = self.player
        _avpvc.showsPlaybackControls = false
        return _avpvc
    }()
    
    private var eventSink: FlutterEventSink?
    
    private var eventChannel: FlutterEventChannel!
    private var methodChannel: FlutterMethodChannel!
    
    lazy private var manager: BCOVPlayerSDKManager = {
        let _manager = BCOVPlayerSDKManager.shared()!
        return _manager
    }()
    
    private var playbackService: BCOVPlaybackService!
    private var playbackController: BCOVPlaybackController!
    
    init(frame:CGRect, viewId: Int64, messenger: FlutterBinaryMessenger, args: Any?) {
        self.viewId = viewId
        
        super.init()
        
        self.avpvc.view.frame = frame
        
        self.setupEventChannel(withViewIdentifier: viewId, messenger: messenger, instance: self)
        self.setupMethodChannel(withViewIdentifier: viewId, messenger: messenger, instance: self)
        
        let rootController = UIApplication.shared.delegate?.window??.rootViewController
        rootController?.addChild(self.avpvc)
        
        /* data as JSON */
        let parsedData = args as! [String: Any]
        
        let playbackControllerArgs = parsedData["playbackController"] as! [String: Any]
        let autoPlay = playbackControllerArgs["autoPlay"] as! Bool
        let autoAdvance = playbackControllerArgs["autoAdvance"] as! Bool
        self.playbackController = self.manager.createPlaybackController()!
        self.playbackController.delegate = self
        self.playbackController.isAutoPlay = autoPlay
        self.playbackController.isAutoAdvance = autoAdvance
        self.playbackController.options = [kBCOVAVPlayerViewControllerCompatibilityKey: true]
        
        let playbackServiceArgs = parsedData["playbackService"] as! [String: Any]
        let accountId = playbackServiceArgs["accountId"] as! String
        let policyKey = playbackServiceArgs["policyKey"] as! String
        self.playbackService = BCOVPlaybackService(accountId: accountId, policyKey: policyKey)
    }
    
    private func setupEventChannel(withViewIdentifier viewId: Int64, messenger:FlutterBinaryMessenger, instance: BCOVVideoPlayer) {
        // register for Flutter event channel
        let eventChannelName = String(format: "bcov.flutter/event_channel_%lld", viewId)
        instance.eventChannel = FlutterEventChannel(name: eventChannelName, binaryMessenger: messenger, codec: FlutterJSONMethodCodec.sharedInstance())
        instance.eventChannel.setStreamHandler(instance)
    }
    
    private func setupMethodChannel(withViewIdentifier viewId: Int64, messenger:FlutterBinaryMessenger, instance: BCOVVideoPlayer) {
        // register for Flutter method channel
        let methodChannelName = String(format: "bcov.flutter/method_channel_%lld", viewId)
        instance.methodChannel = FlutterMethodChannel(name: methodChannelName, binaryMessenger: messenger)
        instance.methodChannel.setMethodCallHandler({
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            self?.handle(call, result: result)
        })
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if "setVideo" == call.method {
            /* data as JSON */
            let parsedData = call.arguments as! [String: Any]
            
            let videoId = parsedData["videoId"] as! String
            let authToken = parsedData["authToken"] as? String
            let parameters = parsedData["parameters"] as? [AnyHashable : Any]
            
            var configuration = [kBCOVPlaybackServiceConfigurationKeyAssetID:videoId]
            if authToken != nil {
                configuration[kBCOVPlaybackServiceConfigurationKeyAuthToken] = authToken
            }
            playbackService?.findVideo(withConfiguration: configuration, queryParameters: parameters, completion: { [weak self] (video: BCOVVideo?, jsonResponse: [AnyHashable: Any]?, error: Error?) in
                
                if let video = video {
                    self?.playbackController.setVideos([video] as NSFastEnumeration)
                }
                else {
                    self?.eventSink?(["name": "onError"])
                }
                result(true)
            })
        }
        else if "play" == call.method {
            self.playbackController.play()
        }
        else if "pause" == call.method {
            self.playbackController.pause()
        }
        else {
            result(FlutterMethodNotImplemented)
        }
    }
}

extension BCOVVideoPlayer: FlutterPlatformView {
    
    func view() -> UIView {
        return self.avpvc.view
    }
}

extension BCOVVideoPlayer: FlutterStreamHandler {
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

extension BCOVVideoPlayer: BCOVPlaybackControllerDelegate {
    
    func playbackController(_ controller: BCOVPlaybackController!, didAdvanceTo session: BCOVPlaybackSession!) {
        self.player = session.player
        self.avpvc.player = session.player
        
        let duration = session.video.properties["duration"] as! Double
        self.eventSink?(["name": "didAdvanceToPlaybackSession", "duration": duration, "isPlaying": self.playbackController.isAutoPlay])
    }
    
    func playbackController(_ controller: BCOVPlaybackController!, playbackSession session: BCOVPlaybackSession!, didProgressTo progress: TimeInterval) {
        if !progress.isInfinite {
            self.eventSink?(["name": "didProgressTo", "progress": (progress * 1000)])
        }
    }
}


class BCOVVideoPlayerFactory: NSObject {
    
    private var registar: FlutterPluginRegistrar
    private var messenger: FlutterBinaryMessenger
    
    init(registrar: FlutterPluginRegistrar) {
        self.registar = registrar
        self.messenger = registrar.messenger()
        super.init()
    }
    
}

extension BCOVVideoPlayerFactory: FlutterPlatformViewFactory {
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        return BCOVVideoPlayer(frame: frame, viewId: viewId, messenger: self.messenger, args: args)
    }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}
